File: unit_1_assignment.py
# ChedChichester_unit1_BIOL672.py
# Name: Ched Chichester
# Operating System: Windows 11
# Context: Unit 1 Assignment Part 2
# Date Created: September 28, 2022
# Date of Last Revision: October 19, 2022
# Packages: math, matplotlib, subprocess, pandas, sklearn, distfit, factor_analyzer
# Data files: iris_tab.txt, iris_size_tab.txt

import subprocess as sub  # Import subprocess to call R function
import pandas as pd  # Imports pandas for creating dataframes
import seaborn as sea  # Import seaborn for plotting sca
from math import log  # Import log to calculate BIC
from distfit import distfit  # Import distfit for distribution fitting
from matplotlib import pyplot as plt  # Imports pyplot for plotting
from sklearn.datasets import load_wine  # Import datasets for analysis
from sklearn.preprocessing import StandardScaler  # Import StandardScaler to standardize values for PCA
from sklearn.decomposition import PCA  # Import PCA to return principal components of data
from sklearn.cluster import KMeans  # Import KMeans to perform k-means clustering
from sklearn import mixture  # Import mixture for GMM model
from factor_analyzer import FactorAnalyzer  # Import FactorAnalyzer to perform factor analysis

# I wanted to execute my R script from this Python script. The following code seems to technically execute my R script,
# but it seems to have an issue with the portion of my R code that writes to .txt files and prints to pdf files (at
# least when I run it on my Windows 11 PC). It does print anything that I would print to my R console to my Python
# console, but I printed all of my R data to output files so nothing will be output to my Python window from this
# except for the "Running R Script..." and "Done running R script!"
sub.call(["C:\\Program Files\\R\\R-4.2.1\\bin\\Rscript", "--vanilla",
          "C:\\Users\\ched1\\RStudio Projects\\unit_1_assignment.r"])  # Run R code via Rscript and denote filepath

# Initialize variables for analysis in downstream-defined functions
iris_size_data = pd.read_csv("iris_size_tab.txt", delimiter="\t")  # Read in iris_size_tab.txt data
iris_size_data = iris_size_data.drop(columns=["species", "avg_size"])  # Drop the species column
sepal_length = iris_size_data.sepal_length  # Assign sepal_length factor
sepal_width = iris_size_data.sepal_width  # Assign sepal_width factor
petal_length = iris_size_data.petal_length  # Assign petal_length factor
petal_width = iris_size_data.petal_width  # Assign petal_width factor
iris_features = [sepal_length, sepal_width, petal_length, petal_width]  # Create array of features for analysis
iris_size_data_standardized = StandardScaler().fit_transform(iris_features)  # Standardize data before PCA and FA


def execute_pca():  # Define function for PCA

    # Create PCA object and execute PCA to obtain loadings and dataframe
    pca = PCA(n_components=4)  # Initialized PCA object with four dimensions
    pca_data = pca.fit_transform(iris_size_data_standardized)  # Execute PCA on data with 4 dimensions
    pca_df = pd.DataFrame(data=pca_data,
                          columns=["PC 1", "PC 2", "PC 3", "PC 4"])  # Create dataframe with principal components
    print("\nPCA Loadings: \n", pca_df)  # Print PCA loadings for analysis
    explained_variance_ratio = pca.explained_variance_ratio_  # Assign the explained variance ratio
    print("\nExplained variance ratios:" + "\nPC 1: " + str(explained_variance_ratio[0]) + "\nPC 2: " +
          str(explained_variance_ratio[1]) + "\nPC 3: " +
          str(explained_variance_ratio[2]) + "\nPC 4: " +
          str(explained_variance_ratio[3]))  # Print explained variance ratio for analysis

    # Plot Explained Variance Ratio figure
    plt.figure(figsize=(6, 6))  # Creates figure with specified size
    plt.bar(range(4), explained_variance_ratio, align="center", label=' Individual Explained Variance', color="blue")
    # Plots bar graph to show how much explained variance is attributed to each principal component
    plt.ylabel('Explained Variance Ratio')  # Label y axis
    plt.xlabel('Principal Components')  # Label x axis
    plt.title("PCA Scree Plot")  # Title plot
    plt.legend(loc='best')  # Create legend in ideal location
    plt.tight_layout()  # Fits labels within figure window
    plt.show()  # Show plot
    print(
        "\nThe data reduction was successful. As shown by the scree plot, only three principal components are required"
        "to explain the variance in the data. This means that the data can be adequately analyzed in three dimensions"
        " without \nlosing any variability. Furthermore, as can be seen visually from the scree plot and "
        "quantitatively from the explained variance ratios, PC 1 and PC 2 account for about 99.7% of the total "
        "variance. Therefore, one could \narguably reduce the dimensions down from four to two and only sacrifice a"
        "very small amount of variability in the data. With regards to the principal component loadings, none of them"
        "had all positive or negative values \n(with the exception of PC 4, which technically had all positive "
        "values, which were effectively 0) and therefore the principal components were all constructed with some "
        "combination of positive and negative combinations of \nsepal_length, sepal_width, petal_length, and "
        "petal_width. It is difficult to say any of the principal components captures much more size variation than "
        "the others because the average loading value for each principal \ncomponent is 0. So, each component is "
        "taking equal positive and negative weights of the features. However, it is worth noting that PC 1 has the "
        "largest weighting values, and therefore it could be said PC 1 is taking the \nhighest percentage of the "
        "features into account. PC 1 is particularly strongly driven by sepal_length and petal_width while PC 2 is "
        "strongly driven by sepal_width and petal_length.")  # Verbal interpretation of PCA results


def execute_factor_analysis():  # Define function for factor analysis
    factor_analysis = FactorAnalyzer(n_factors=3, rotation="varimax")  # Prepare factor analysis model with 3 factors
    factor_analysis.fit(iris_size_data)  # Fit factor analysis model to data
    ev, v = factor_analysis.get_eigenvalues()  # Return eigenvalues and eigenvectors
    print("\nFactor Analysis Eigenvalues: " + str(ev))  # Print eigenvalues
    print("\nFactor Analysis Loadings: \n" + str(factor_analysis.loadings_))  # Print FA loadings for analysis
    print("\nLooking at the factor analysis loadings, factor 1 is strongly associated with the sepal_length,"
          " petal_length, and petal_width. Factor 2 is strongly associated with the sepal_width."
          "By analyzing the eigenvalues from the factor \nanalysis of the covariance matrix, only one eigenvalue is "
          "greater than 1, with a second eigenvalue being nearly 1. Thus, factor 1 is the most significant factor"
          "with factor 2 potentially being somewhat significant as well. \nFactor 3 is not very significant. The main"
          "characteristics that cluster together are sepal_length, petal_length, and petal_width as can be seen in "
          "factor 1. Also, petal_length and petal_width cluster together as both factor 2 \nand factor 3 have values "
          "for these variables that are similar to one another. If two traits are far apart along the axis of a "
          "significant factor, this signifies the traits are not correlated or even inversely correlated, and \nvice "
          "versa. Yes, this factor analysis was successful in identifying two latent factors that signify relationships"
          " between the variables within the model.")  # Verbal interpretation of factor analysis


def execute_kmeans():  # Define function for k-means clustering
    iris_data = pd.read_csv("iris_tab.txt", delimiter="\t")  # Read in iris_tab.txt data
    sea.lmplot(x="petal_length", y="petal_width", data=iris_data, hue="species", fit_reg=False)  # Plot scatter plot of
    # petal_length vs petal_width without regression fit lines and select point color by species
    plt.title("Species-Categorized Petal Length vs Petal Width")  # Title plot
    plt.xlabel("Petal Length")  # Labels x axis
    plt.ylabel("Petal Width")  # Labels y axis
    plt.tight_layout()  # Fits labels within figure window
    plt.show()  # Show plot

    k_means = KMeans(n_clusters=2)  # Specify the number of clusters to predict via k-means clustering algorithm
    k_means_cluster = k_means.fit_predict(iris_data[["petal_length", "petal_width"]])  # Predict k-means clusters for
    # petal_length and petal_width
    iris_data["cluster"] = k_means_cluster  # Create new column to match each row with its assigned cluster
    iris_df_1 = iris_data[iris_data.cluster == 0]  # Create new dataframe for the first cluster
    iris_df_2 = iris_data[iris_data.cluster == 1]  # Create new dataframe for teh second cluster
    plt.scatter(iris_df_1.petal_length, iris_df_1.petal_width, color="red")  # Plot scatter of first cluster red
    plt.scatter(iris_df_2.petal_length, iris_df_2.petal_width, color="blue")  # Plot scatter of second cluster blue
    plt.xlabel("Petal Length")  # Label x axis
    plt.ylabel("Petal Width")  # Label y axis
    plt.title("K-Means-Clustered Petal Length vs Petal Width")  # Title plot
    plt.tight_layout()  # Fits labels within figure window
    plt.show()  # Show plot


def multi_model_inference():  # Define function for multi-model inference
    wine_data = load_wine()  # Load wine dataset
    wine_dataframe = pd.DataFrame(wine_data.data)  # Creates dataframe of the wine dataset
    wine_alcohol_content = wine_dataframe.iloc[:, 0]  # Extracts first column of dataframe, which represents alcohol
    # content
    n = len(wine_alcohol_content)  # Calculate n for BIC calculation
    k = 1  # Assign k for BIC calculation

    dist_norm = distfit(distr="norm")  # Creates normal distribution fit object
    dist_norm.fit_transform(wine_alcohol_content)  # Fits normal distribution to data
    dist_norm.plot(title="Normal Distribution Fitting", figsize=(8, 8))  # Plots normal distribution fit
    plt.show()  # Show plot
    norm_rss = 1.21089  # Assign normal distribution RSS to calculate BIC
    norm_bic = n * log(norm_rss/n) + k * log(n)  # Calculate normal distribution BIC

    dist_lognorm = distfit(distr="lognorm")  # Creates log-normal distribution fit object
    dist_lognorm.fit_transform(wine_alcohol_content)  # Fits log-normal distribution to data
    dist_lognorm.plot(title="Log-Normal Distribution Fitting", figsize=(8, 8))  # Plots log-normal distribution fit
    plt.show()  # Show plot
    lognorm_rss = 1.21219  # Assign normal distribution RSS to calculate BIC
    lognorm_bic = n * log(lognorm_rss/n) + k * log(n)  # Calculate log-normal distribution BIC

    dist_expon = distfit(distr="expon")  # Creates exponential distribution fit object
    dist_expon.fit_transform(wine_alcohol_content)  # Fits exponential distribution to data
    dist_expon.plot(title="Exponential Distribution Fitting", figsize=(8, 8))  # Plots exponential distribution fit
    plt.show()  # Show plot
    expon_rss = 3.32596  # Assign normal distribution RSS to calculate BIC
    expon_bic = n * log(expon_rss/n) + k * log(n)  # Calculate exponential distribution BIC

    dist_all = distfit()  # Creates fit object without defining distribution to compare each method
    dist_all.fit_transform(wine_alcohol_content)  # Fits data to several distributions for comparison
    dist_all.plot_summary()  # Return summary of distribution performances
    plt.title("Distribution Fit Summary")  # Titles plot
    plt.show()  # Show plot

    wine_alcohol_content = wine_alcohol_content.values
    wine_alcohol_content = wine_alcohol_content.reshape(-1, 1)
    gmm = mixture.GaussianMixture(n_components=3)  # Creates GMM object with 3 classes/latent variables
    gmm.fit(wine_alcohol_content)  # Fit GMM to model
    gmm_bic = gmm.bic(wine_alcohol_content)  # Calculate BIC on GMM fit

    print("\nThe normal distribution BIC is :" + str(norm_bic))  # Print normal distribution BIC
    print("\nThe log-normal distribution BIC is :" + str(lognorm_bic))  # Print log-normal distribution BIC
    print("\nThe exponential distribution BIC is :" + str(expon_bic))  # Print exponential distribution BIC
    print("\nThe GMM BIC is: " + str(gmm_bic))  # Print GMM BIC
    print("THe result of my model fitting and BIC analysis shows that the log-normal distribution is the best fit for"
          " the data. Thus, this model testing did not indicate the presence of latency. Interestingly, the BIC values"
          "\nI calculated for the normal, log-normal, and exponential distributions are all negative while the GMM BIC "
          "is positive. I calculated the BIC for the non-GMM distributions utilizing the RSS from their fits, but "
          "perhaps there\nis an issue or discrepancy between this method and the method from the GMM module, which had"
          " a built-in method to return the BIC. It's also worth mentioning that the BIC values for the normal, "
          "log-normal, and exponential\ndistributions align with the RSS values in that the ones with lower RSS values"
          "have lower BIC values, indicated a better fit. Unfortunately, I had some trouble producing a plot of the"
          " GMM distribution, so simply assessing\nthe BIC values is all I am able to do for now.")
    # Print verbal interpretation


def main():  # Execute defined functions
    execute_pca()  # Execute PCA
    execute_factor_analysis()  # Execute factor analysis
    execute_kmeans()  # Execute k-means clustering
    multi_model_inference()  # Execute multi-model inference


if __name__ == "__main__":  # Execute Python functions
    main()  # Execute main function above

File: unit_1_assignment.r
# ChedChichester_unit1_BIOL672.r
# Name: Ched Chichester
# Operating System: Windows 11
# Context: Unit 1 Assignment Part 1
# Date Created: September 21, 2022
# Date of Last Revision: October 19, 2022
# Libraries: ggplot2, dplyr
# Data files: mtcars, diamonds, iris_tab, iris_purchase, iris_tab_bgnoise, 
# iris_tab_missing, iris_tab_randclass

print("Running R script...") # Let the user know the R script is now running

library(ggplot2) # Loads the ggplot2 library for ggplot plotting features
library(dplyr) # Loads the dplyr library to group dataframe elements

# The following code will execute #1, #2, and #3 of Unit Assignment 1
sink("desc.txt") # Opens link to desc.txt
data_set <- rnorm(5000) # Creates a random data set of 
# 5000 normally distributed numbers
sample_mean <- mean(data_set) # Calculates the mean of the data set
standard_deviation <- sd(data_set) # Calculates the standard deviation of the 
# data set
normal_data_sequence <- seq(min(data_set), max(data_set), length = 50) # Creates
# a sequence of values
normal_data <- dnorm(normal_data_sequence, mean = mean(data_set),
                     sd = sd(data_set)) # Creates a normally distributed 
# distribution
sample_mean_string <- sprintf("The sample mean is %s", sample_mean) # Creates 
# string of sample mean
standard_deviation_string<- sprintf("The sample standard deviation is %s",
                                    standard_deviation) # Creates string of 
# standard deviation
print(sample_mean_string) # Print Sample Mean
print(standard_deviation_string) # Print standard deviation
sink() # Close sink connection

pdf("histo.pdf")
plot_1 <- hist(data_set, prob = TRUE, main = 
                 "Histogram With Density Curve and Normal Curve") # Titles plot
lines(density(data_set), col = 4, lwd = 2) # Plots distribution curve
lines(normal_data_sequence, normal_data, col = 2, lwd = 2) # Plots normal curve
dev.off() # Closes link to PDF file

# The following code will execute #4 of Unit Assignment 1
write.csv(mtcars,"ANOVA_Test_Data") # Writes mtcars built-in R data to a CSV 
# file (not necessary to do, but this is just to practice writing/ reading data)
anova_data <- read.csv("ANOVA_Test_Data", sep = ",") # Read in ANOVA data
cyl <- anova_data$cyl # Extracts cyl column from data
hp <- anova_data$hp # Extracts hp column from data
anova_test <- oneway.test(hp~cyl) # Conducts one-way ANOVA
anova_dataframe <- data.frame(cyl, hp) %>% # Creates dataframe with cyl and hp
  group_by(cyl) %>% # Groups dataframe by cyl
  summarise(mean_hp = mean(hp), sd_hp = sd(hp), count = n(), se_hp = 
              (sd_hp/sqrt(count))) # Summarizes statistics for each group

plot_2 <- ggplot(anova_dataframe,
                 aes(x = cyl, y = mean_hp,fill = factor(cyl))) +
  geom_bar(stat = "identity", position = position_dodge()) + 
  labs(x = "Number of Cylinders", y = "Horesepower", 
       title = "mtcars Cylinders vs Horsepower") + 
  theme(plot.title = element_text(hjust =0.5)) + 
  geom_errorbar(aes(ymin = mean_hp-se_hp, ymax = mean_hp+se_hp)) # Plots bar
# chart with error bars for each cyl group and adds labels/titles accordingly

pdf("Error_Bar_Chart.pdf")
print(plot_2) # Prints the plot
dev.off() # Close link to PDF file

anova_data <- data.frame(cyl, hp)
split_anova_data <- split(anova_data, anova_data$cyl) # Splits data by cyl value
split_anova_data_4 <- unlist(split_anova_data$`4`) # Changes data type from list
split_anova_data_6 <- unlist(split_anova_data$`6`) # Changes data type from list
split_anova_data_8 <- unlist(split_anova_data$`8`) # Changes data type from list

# Execute t tests with Benjamini-Hochberg correction
t_test_1 <- t.test(split_anova_data_4, split_anova_data_6, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "hochberg")
t_test_2 <- t.test(split_anova_data_4, split_anova_data_8, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "hochberg")
t_test_3 <- t.test(split_anova_data_6, split_anova_data_8, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "hochberg") 

# Execute t tests with Bonferroni correction
t_test_4 <- t.test(split_anova_data_4, split_anova_data_6, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "bonferroni")
t_test_5 <- t.test(split_anova_data_4, split_anova_data_8, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "bonferroni")
t_test_6 <- t.test(split_anova_data_6, split_anova_data_8, mu = 0, 
                   alt = "two.sided", paired = FALSE, var.equal = FALSE, 
                   conf.level = 0.95, p.adjust.method = "bonferroni") 

sink("ANOVA_test_results.txt") # Opens sink to print to .txt file
print("ANOVA test results for mtcars of cyl and hp:")
print(anova_test) # Prints ANOVA results
print("t_test results of 4 cyl and 6 cyl with Benjamini-Hochberg correction:")
# Prints description of results
print(t_test_1) # Prints t test 1 results
print("t_test results of 4 cyl and 8 cyl with Benjamini-Hochberg correction:")
# Prints description of results
print(t_test_2) # Prints t test 2 results
print("t_test results of 6 cyl and 8 cyl with Benjamini-Hochberg correction:")
# Prints description of results
print(t_test_3) # Prints t test 3 results
print("t_test results of 4 cyl and 6 cyl with Bonferroni correction:")
# Prints description of results
print(t_test_4) # Prints t test 4 results
print("t_test results of 4 cyl and 8 cyl with Bonferroni correction:")
# Prints description of results
print(t_test_5) # Prints t test 5 results
print("t_test results of 6 cyl and 8 cyl with Bonferroni correction:")
# Prints description of results
print(t_test_6) # Prints t test 6 results
print("From the data, we can see that the t test results were exactly the same
      for both theh Bonferroni and Benjamini-Hochberg corrections. Although, 
      this is not typically the case, it may have resulted here due to the 
      rather small sample size.")
sink() # Close sink connection

# The following code will execute #5 of Unit Assignment 1
sink("Non-parametric_Tests.txt")
kw_test <- kruskal.test(hp~cyl) # Performs Kruskal-Wallis test
print(kw_test)
pearson_correlation <- cor(x = anova_data,use = "complete.obs", 
                           method = "pearson") # Calculates Pearson correlation
spearman_correlation <- cor(x = anova_data,use = "complete.obs", 
                           method = "spearman") # Calculates Spearman correlation
print("Pearson Correlation Test:")
print(pearson_correlation) # Print Pearson correlation results
print("Spearman Correlation Test:")
print(spearman_correlation) # Print Spearman correlation results

ks_test <- ks.test(cyl, hp) # Performs Kolmogorov-Smirnov test
print(ks_test) # Prints KS results
plot_3 <- ggplot(anova_data, aes(x = cyl, y = hp, color = factor(cyl))) + 
  geom_point() + labs(x = "Number of Cylinders", y = "Horsepower",  
                      title = "Scatterplot of Cylinders vs Horsepower") + 
theme(plot.title = element_text(hjust =0.5)) # Plot cyl vs hp scatterplot)
pdf("Scatterplot_of_Cylinders_vs_Horsepower.pdf")
print(plot_3) # Print scatterplot
dev.off()

# The following code will execute #6 of Unit Assignment 1
linear_regression <- lm(anova_data) # Returns linear regression slope and 
# coefficient
print("Linear regression results:")
print(linear_regression) # Print linear regression results
print("The assumption of normality did not hold as the p value for the Kolmogorov-Smirnov test was 2.2e-16. This was expected as the data consists of horespower generated from the number of cylinders in an engine, which would be expected to increase as cylinders increase. However, in this case, the results form both the parametric and non-parametric tests found similar results of significant difference between the three groups. This is likely due to the large disparity in the horespower for each cylinder group. If the disparity was not as large, the tests may have differed more in their findings.")
sink() # Close connection to .txt file

pdf("Cylinder vs_Horsepower_Scatterplot.pdf") # Opens pdf to print plot
plot_4 <- ggplot(anova_data, aes(x = cyl, y = hp, color = factor(cyl))) + 
  geom_point() + labs(x = "Number of Cylinders", y = "Horsepower",  
       title = "Scatterplot of Cylinders vs Horsepower") + 
  geom_smooth(aes(x = cyl, y = hp), method = "lm", formula = y~x)
  theme(plot.title = element_text(hjust =0.5)) # Plot cyl vs hp scatterplot with linear regression line
print(plot_4) # Print plot
plot.new() # Open new plot for text command
text(0.5,0.5,"The results are consistent with what was discovered in the 
correlation analysis earlier. The data in this case follows a linear 
relationship where the horsepower variable is very highly positively correlated 
with the cylinders variable. Correlation is better suited for quickly 
determining how strong of a relationship exists between the variables while 
regression is better for representing the model and interpolating or 
     extrapolating data points.")
dev.off() # Close link to scatterplot pdf

# The following code will execute #7 and #8 of Unit Assignment 1
write.csv(diamonds,"Multivariate_Data.txt") # Writes diamonds built-in R data to a CSV 
# file (not necessary to do, but this is just to practice writing/ reading data)
multivariate_data <- read.csv("Multivariate_Data.txt", sep = ",") # Read in data
cut <- multivariate_data$cut
cut_category <- as.numeric(factor(cut, levels = c("Ideal", "Premium","Very Good", "Good", "Fair"))) # Converts strings of cut into numerical values
depth <- multivariate_data$depth
table <- multivariate_data$table
x <- multivariate_data$x
y <- multivariate_data$y
z <- multivariate_data$z
multivariate_dataframe <- data.frame(cut_category, depth, table, x, y, z) # Create dataframe with 5 variables and 1 category (cut)
dependent_vars <- cbind(depth, table, x, y, z) # Combine dependent variables into a dataframe
manova_model <- manova(formula = dependent_vars~cut_category, data = multivariate_data)
sink("MANOVA_results")
print(summary(manova_model, test = "Pillai")) # Prints a summary of a Pillai MANOVA test
print("As can be seen in the MANOVA summary, the p value is extremely small at only 2.2e-16, which I believe is the minimum value that R can report here. This suggests that the cut of the diamond is significantly influenced by the depth, table, x, y, and z variables. This intuitively makes sense as the dimensions of the diamond are what are assessed in determining the quality of the cut.")
sink() # Close connection

# The following code will execute #9 of Unit Assignment 1
sink("Multiple_Regression_Results.txt") # Opens a new file to print multiple regression results
multiple_regression_data <- lm(formula = cut_category ~ depth + table + x + y + z, data = multivariate_dataframe) # Fits
# model of cut_category to other independent variables
within_category_df <- multivariate_dataframe[multivariate_dataframe$cut_category == 1, ]
multiple_regression_data_within_category <- lm(formula = cut_category ~ depth + table + x + y + z, data = within_category_df) # Fits model within cut category
print(summary(multiple_regression_data)) # Prints summary of lm results
print(summary(multiple_regression_data_within_category)) # Prints summary of lm results
print("Per the results of the multiple regression, all of the independent variables in this scenario of depth, table, x, y, and z together have a significant impact on the cut category, but they do not have much of a different impact than one another. This aligns with my expectation that the cut category should be directly related to each of these measures. Thus, similar data may be achieved even if a few of the variables were removed. It is interesting to note that the p values for depth and table are much lower than the p values for x, y, and z. This implies that the depth and table are more influential in determining the cut category than the other independent variables. Therefore, if I were to pursue dimensionality reduction on this data, I might consider neglecting some of the variables, especially the z variable, which contributes the least to cut category.")
sink() # Close connection

# The following code will execute #10 of Unit Assignment 1
sink("ANCOVA_Results.txt") # Open .txt file for ANCOVA results
clarity <- multivariate_data$clarity
clarity_category <- as.numeric(factor(clarity, levels = c("FL", "IF","VVS1", "VVS2", "VS1", "VS2", "SI1", "SI2", "I1", "I2", "I3"))) # Converts strings of clarity into numerical values
price <- multivariate_data$price
ancova_dataframe <- data.frame(price, cut_category, clarity_category)
ancova_data <- aov(formula = price ~ clarity_category*cut_category, data = ancova_dataframe) # Run ANCOVA test
print(summary(ancova_data)) # Print ANCOVA results
print("As expected, the clarity and cut both significantly impact the price of the diamond in their own right. However, when controlling for the clarity covariate, the cut by itself has a somewhat less significant impact on the price. In addition, the interaction of the two independent variables is quite strong with a very low p value found for this interaction." )
sink() # Close connections

# The following code will execute #11 and #12 of Unit Assignment 1
sink("iris_original_and_corrupted_results.txt")
normal_iris <- read.delim("iris_tab.txt", sep = "\t")
sepal_length <- normal_iris$sepal_length
sepal_width <- normal_iris$sepal_width
petal_length <- normal_iris$petal_length
petal_width <- normal_iris$petal_width
species <- as.numeric(factor(normal_iris$species, levels = c("setosa", "versicolor","virginica")))
normal_iris_df <- data.frame(sepal_length, sepal_width, petal_length, petal_width, species)
normal_multiple_regression <- lm(formula = species ~ sepal_length + sepal_width + petal_length + petal_width, data = normal_iris_df)
# Fits multiple regression to normal iris data
print(summary(normal_multiple_regression))

# Repeat for iris_tab_bgnoise:
bgnoise_iris <- read.delim("iris_tab_bgnoise.txt", sep = "\t")
sepal_length <- normal_iris$sepal_length
sepal_width <- normal_iris$sepal_width
petal_length <- normal_iris$petal_length
petal_width <- normal_iris$petal_width
species <- as.numeric(factor(bgnoise_iris$species, levels = c("setosa", "versicolor","virginica")))
bgnoise_iris_df <- data.frame(sepal_length, sepal_width, petal_length, petal_width, species)
bgnoise_multiple_regression <- lm(formula = species ~ sepal_length + sepal_width + petal_length + petal_width, data = bgnoise_iris_df)
# Fits multiple regression to normal iris data
print(summary(bgnoise_multiple_regression))

# Repeat for iris_tab_randclass:
randclass_iris <- read.delim("iris_tab_randclass.txt", sep = "\t")
sepal_length <- normal_iris$sepal_length
sepal_width <- normal_iris$sepal_width
petal_length <- normal_iris$petal_length
petal_width <- normal_iris$petal_width
species <- as.numeric(factor(randclass_iris$species, levels = c("setosa", "versicolor","virginica")))
randclass_iris_df <- data.frame(sepal_length, sepal_width, petal_length, petal_width, species)
randclass_multiple_regression <- lm(formula = species ~ sepal_length + sepal_width + petal_length + petal_width, data = randclass_iris_df)
# Fits multiple regression to normal iris data
print(summary(randclass_multiple_regression))

# Repeat for iris_tab_missing:
missing_iris <- read.delim("iris_tab_missing.txt", sep = "\t")
sepal_length <- normal_iris$sepal_length
sepal_width <- normal_iris$sepal_width
petal_length <- normal_iris$petal_length
petal_width <- normal_iris$petal_width
species <- as.numeric(factor(missing_iris$species, levels = c("setosa", "versicolor","virginica")))
missing_iris_df <- data.frame(sepal_length, sepal_width, petal_length, petal_width, species)
missing_multiple_regression <- lm(formula = species ~ sepal_length + sepal_width + petal_length + petal_width, data = missing_iris_df)
# Fits multiple regression to normal iris data
print(summary(missing_multiple_regression))

# Print interpretation of normal_iris vs corrupted iris data
print("It's interesting to note that the multiple regression results are exactly the same for the original iris dataset as the corrupted data sets with added noise or with missing values. The only corrupted data set I analyzed that returned different results was the random classified data. It appears the bgnoise and missing data sets have no impact on the influence of the factors on the dependent variable. However, the randomly classified data set greatly reduces the ability of any factor to predict the classification. Though, even in the randomly classified data set, the combination of all four factors is stil able to predict the classification fairly well with a p value of ~0.0003. Although, we cannot be certain the classification being predicted is correct (due to it being randomly classified) without further analysis of the data.")
sink() # Close connection

# The following code will execute #12 of Unit Assignment 1
sink("iris_purchase_results.txt")
iris_purchase <- read.csv("iris_purchase.txt", sep = "\t")
color <- as.numeric(factor(iris_purchase$color, levels = c("yellow", "orange", "red", "blue")))
attractiveness <- iris_purchase$attractiveness
review <- iris_purchase$review
likelytobuy <- iris_purchase$likelytobuy
sold <- as.numeric(factor(iris_purchase$sold, levels = c("TRUE", "FALSE")))
iris_purchase_df <- data.frame(color, attractiveness, review, likelytobuy, sold)
iris_purchase_multiple_regression <- lm(sold ~ color + attractiveness + review + likelytobuy)
print(summary(iris_purchase_multiple_regression))
print("Based on the results of the multiple regression comparing the influence of color, attractiveness, review, and likelytobuy on if a flower is sold, it can be seen that no individual factor has a significant impact on selling the flower. Although, all factors together have a strong predictability on if the flower is sold. The most significant factor on if a flower sells is the likelytobuy factor, which intuitively makes sense because this factor should aim to capture the likelihood that a flower is sold to a certain customer.")
sink() # Close connection

print("Done running R script") # Let the user know the R script is done running
